src/fns.ts

```diff
@@ -50,23 +64,100 @@ export const rateLimit = <T = unknown>(limit: number) => {
    * Wait for all promises to resolve
    */
   const finish = async () => {
-    debug('finish')
+    rl('finish')
     await Promise.all(set)
   }
   return { add, finish }
 }
 
 /**
+ * Limit throughput by sleeping until the rate (items/sec)
+ * is less than or equal to `maxItemsPerSec`.
  *
+ * Example:
+ * ```typescript
+ * const limiter = throughputLimiter(1000)
+ *
+ * for(const batch of batches) {
+ *   // Will wait until the rate is <= `maxItemsPerSec`
+ *   await limiter.throttle(batch.length)
+ *   console.log('Items/sec %d', limiter.getCurrentRate())
+ * }
+ * ```
+ */
+export const throughputLimiter = (
+  maxItemsPerSec: number,
+  options: ThroughputLimiterOptions = {}
+) => {
+  const slidingWindow: { startTime: number; numItems: number }[] = []
+  const windowLength = options.windowLength || 3
+  const sleepTime = options.sleepTime || 100
+  tl('init - maxItemsPerSec %d', maxItemsPerSec)
+  tl('init - windowLength %d', windowLength)
+  tl('init - sleepTime %d', sleepTime)
+
+  /**
+   * Get the current rate (items/sec). Returns 0 if `throttle` has not been
+   * called.
+   */
+  const getCurrentRate = () => {
+    tl('getCurrentRate called')
+    if (slidingWindow.length > 0) {
+      const { startTime } = slidingWindow[0]
+      const numItems = sumBy(slidingWindow, 'numItems')
+      tl('total items %d', numItems)
+      const rate = numItems / ((new Date().getTime() - startTime) / 1000)
+      tl('current rate %d', rate)
+      return rate
+    }
+    tl('current rate 0')
+    return 0
+  }
+
+  /**
+   * Call before processing a batch of items. After the first call, a subsequent
+   * call assumes that the `numItems` from the previous call were processed. A
+   * call to `throttle` may sleep for a given period of time depending on
+   * `maxItemsPerSec` and the total number of items over the current window.
+   */
+  const throttle = async (numItems: number) => {
+    tl('throttle called - %d', numItems)
+    // Skip check if maxItemsPerSec is Infinity
+    if (maxItemsPerSec === Infinity) {
+      tl('exiting throttle - maxItemsPerSec is Infinity')
+      return
+    }
+    while (getCurrentRate() > maxItemsPerSec) {
```

Comment:
It would be nice to add some comments here (and/or getCurrentRate) to explain this mechanism and why this causes the execution to sleep

Reply:
Added.